### PR TITLE
feature/get-all-posts

### PR DESCRIPTION
**Pull request by Belal-Alkhatib**

### CODE REVIEW

File: app/src/main/java/com/cupcake/jobsfinder/data/remote/modle/PostDto.kt
Comment (AmeerAmjed) [inline app/src/main/java/com/cupcake/jobsfinder/data/remote/modle/PostDto.kt:1]: this all field no required not need yet keep just required just id and content and createAt

File: app/src/main/java/com/cupcake/jobsfinder/ui/posts/PostUIState.kt
Comment (AmeerAmjed) [inline app/src/main/java/com/cupcake/jobsfinder/ui/posts/PostUIState.kt:1]: keep just required

**Review by AmeerAmjed — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/cupcake/jobsfinder/domain/mapper/PostMapper.kt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package com.cupcake.jobsfinder.domain.mapper
+
+import com.cupcake.jobsfinder.data.remote.modle.PostDto
+import com.cupcake.jobsfinder.domain.model.Post
+import com.cupcake.jobsfinder.mapper.Mapper
+import javax.inject.Inject
+
+class PostMapper @Inject constructor() : Mapper<PostDto, Post> {
```
Comment (msalarab) [inline app/src/main/java/com/cupcake/jobsfinder/domain/mapper/PostMapper.kt]: good job but we are using the extension function.

File: app/src/main/java/com/cupcake/jobsfinder/ui/posts/PostsViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    private val _uiState = MutableStateFlow(PostsUIState())
+    val uiState: StateFlow<PostsUIState> = _uiState
+    init {
+        onGetPosts()
+    }
+
+    private fun onGetPosts(){
+        _uiState.update { it.copy(isLoading = true) }
+
+        viewModelScope.launch(Dispatchers.IO) {
```
Comment (msalarab) [inline app/src/main/java/com/cupcake/jobsfinder/ui/posts/PostsViewModel.kt]: if the result showing on ui u need to use main thread

File: app/src/main/java/com/cupcake/jobsfinder/ui/posts/PostsViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            try {
+                val result = getPostsUseCase()
+                _uiState.update {
+                    it.copy(isLoading = false, postsResult = result.map { post ->
+                        post.toUIState()
+                    })
+                }
+            } catch (e: Exception) {
+                _uiState.update {
+                    it.copy(errors = listOf(ErrorUIState(message = e.message.toString())))
```
Comment (msalarab) [inline app/src/main/java/com/cupcake/jobsfinder/ui/posts/PostsViewModel.kt]: you can create an error list in the global scope to use for all view model.

**Review by msalarab — CHANGES_REQUESTED** [review]
good job.

File: app/src/main/java/com/cupcake/jobsfinder/data/remote/modle/PostDto.kt
Comment (Belal-Alkhatib) [reply app/src/main/java/com/cupcake/jobsfinder/data/remote/modle/PostDto.kt:1]: we already do that in Post class in a domin layer

File: app/src/main/java/com/cupcake/jobsfinder/ui/posts/PostsViewModel.kt (hunk 2 above)
Comment (Belal-Alkhatib) [reply app/src/main/java/com/cupcake/jobsfinder/ui/posts/PostsViewModel.kt]: thanks mustafa, but I don't directly observe the onGetPosts function. I observe the uiState in the UI layer. However, when observing the uiState, I switch the thread to the main thread in the UI layer

File: app/src/main/java/com/cupcake/jobsfinder/domain/mapper/PostMapper.kt (hunk 1 above)
Comment (Belal-Alkhatib) [reply app/src/main/java/com/cupcake/jobsfinder/domain/mapper/PostMapper.kt]: so why the Mapper class are exist 😅

**Review by AmeerAmjed — APPROVED** [review]
good job

File: app/src/main/java/com/cupcake/jobsfinder/data/remote/modle/PostDto.kt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.cupcake.jobsfinder.data.remote.modle
+
+data class PostDto(
```
Comment (TarekIdrees) [inline app/src/main/java/com/cupcake/jobsfinder/data/remote/modle/PostDto.kt:3]: I think these fields may return null in some cases from the API

**Review by TarekIdrees — APPROVED** [review]

**Review by msalarab — APPROVED** [review]

**Pull request merged**